# Student's Score

Aim: What will be predicted score if a student studies for 9.25 hrs/ day?

Notebook used: Jupyter Notbook

### Importing the libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Importing the dataset

In [2]:
dataset = pd.read_csv(r"C:\Desktop\Quick Access\Student Data.csv")
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:,1].values

### Cheking null values

In [3]:
DF = pd.DataFrame(dataset)
DF.shape
DF.isna()

Hours  Scores
0   False   False
1   False   False
2   False   False
3   False   False
4   False   False
5   False   False
6   False   False
7   False   False
8   False   False
9   False   False
10  False   False
11  False   False
12  False   False
13  False   False
14  False   False
15  False   False
16  False   False
17  False   False
18  False   False
19  False   False
20  False   False
21  False   False
22  False   False
23  False   False
24  False   False

### Splitting the model

In [4]:
from sklearn.model_selection import train_test_split
x_train,x_test ,y_train, y_test = train_test_split(x,y,test_size = 0.25,random_state = 0)

### Training the model

In [5]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(x_train, y_train)

LinearRegression()

### Comparing the pedicted score with the actual score

In [6]:
y_pred = regressor.predict(x_test)
df = pd.DataFrame({'Actual Score':y_test,'Predicted_score':y_pred})
df

Actual Score  Predicted_score
0            20        16.844722
1            27        33.745575
2            69        75.500624
3            30        26.786400
4            62        60.588106
5            35        39.710582
6            24        20.821393

### Predicting the score for 9.25 hours of study

In [7]:
print(regressor.predict([[9.25]]))

[93.89272889]


### Coeficient and Intercepts

In [8]:
print(regressor.coef_)
print(regressor.intercept_)

[9.94167834]
1.932204253151646


### Evaluating the result of the model

In [9]:
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error,explained_variance_score
r2score_lin = r2_score(y_test,y_pred)
MSE_lin = mean_squared_error(y_test,y_pred)
MAE_lin = mean_absolute_error(y_test,y_pred)
EVS_lin = explained_variance_score(y_test,y_pred)

### Comparing with Polynomial regression model

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
dataset = pd.read_csv(r"C:\Desktop\Quick Access\Student Data.csv")
x = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values
DF = pd.DataFrame(dataset)
DF.shape
DF.isna()
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.25,random_state = 0)
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
poly_reg = PolynomialFeatures(degree = 4)
regressor = poly_reg.fit_transform(x_train)
lin_reg = LinearRegression()
lin_reg.fit(regressor,y_train)
y_pred = lin_reg.predict(poly_reg.fit_transform(x_test))
df = pd.DataFrame({'Actual Score':y_test,'Predicted_score':y_pred})
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error
r2score_poly = r2_score(y_test,y_pred)
MSE_poly = mean_squared_error(y_test,y_pred)
MAE_poly = mean_absolute_error(y_test,y_pred)
EVS_poly = explained_variance_score(y_test,y_pred)

### Polynomial Regression V/s Linear Regression

In [11]:
error_df = pd.DataFrame({"MSE":[MSE_poly,MSE_lin],"MAE":[MAE_poly,MAE_lin],"EVS":[EVS_poly,EVS_lin],'R2_score':[r2score_poly,r2score_lin]})
error_df.rename(index = {0:'Poly',1:'Lin'},inplace = True)
print('MSE = Mean Squared Error')
print('MAE = Mean absolute Error')
print('EVS = Explained Variance Square')
error_df.T

MSE = Mean Squared Error
MAE = Mean absolute Error
EVS = Explained Variance Square


Poly        Lin
MSE       24.074281  20.332924
MAE        4.083907   4.130880
EVS        0.929441   0.939874
R2_score   0.925131   0.936766

## Conclusion

From the above table it is clear that linear regression model is more accurate than Polynomial regression model therefore as predicted by our Linear reggression model the student will score "93.89272889" marks if he/she studies for 9.25hrs/day

Thank you!